File: solutions/ch13-checkpoint-solution.js
/**
 * solutions/ch13-checkpoint-solution.js
 * 
 * Agent Frida — Checkpoint Chapitre 13
 * Logue tous les appels à send() avec leurs buffers,
 * et intercepte connect() pour contextualiser les connexions.
 *
 * Usage CLI :
 *   frida -f ./client_O0 -l ch13-checkpoint-solution.js --no-pause
 *
 * Usage Python :
 *   Voir ch13-checkpoint-solution.py (client d'orchestration)
 */
'use strict';

// ═══════════════════════════════════════════════════════════
// CONFIGURATION
// ═══════════════════════════════════════════════════════════

const CONFIG = {
    // Filtrage par file descriptor (-1 = désactivé, tout capturer)
    filterFd: -1,

    // Taille maximale du hexdump affiché en console
    hexdumpMaxBytes: 128,

    // Hooker aussi write() sur les fd de sockets connus
    hookWrite: true,
};

// ═══════════════════════════════════════════════════════════
// ÉTAT GLOBAL
// ═══════════════════════════════════════════════════════════

const startTime = Date.now();
let sequenceNumber = 0;

// Table des connexions connues : fd → { ip, port }
const connections = new Map();

// Table des fd de sockets connus (pour filtrer write() sur sockets)
const knownSocketFds = new Set();

// ═══════════════════════════════════════════════════════════
// UTILITAIRES
// ═══════════════════════════════════════════════════════════

/**
 * Retourne le nombre de millisecondes écoulées depuis le
 * début du traçage.
 */
function elapsed() {
    return Date.now() - startTime;
}

/**
 * Parse une structure sockaddr_in pointée par `addrPtr`.
 * Retourne { family, ip, port } ou null si la famille
 * n'est pas AF_INET (2).
 */
function parseSockaddrIn(addrPtr) {
    try {
        const family = addrPtr.readU16();
        if (family !== 2) {  // AF_INET
            return { family, ip: null, port: null };
        }

        // sin_port : 2 octets en network byte order (big-endian)
        const portHi = addrPtr.add(2).readU8();
        const portLo = addrPtr.add(3).readU8();
        const port = (portHi << 8) | portLo;

        // sin_addr : 4 octets
        const ip = [
            addrPtr.add(4).readU8(),
            addrPtr.add(5).readU8(),
            addrPtr.add(6).readU8(),
            addrPtr.add(7).readU8()
        ].join('.');

        return { family, ip, port };
    } catch (e) {
        return null;
    }
}

/**
 * Vérifie si un fd doit être loggué selon la configuration
 * de filtrage.
 */
function shouldLog(fd) {
    if (CONFIG.filterFd === -1) return true;
    return fd === CONFIG.filterFd;
}

// ═══════════════════════════════════════════════════════════
// HOOK : connect()
// ═══════════════════════════════════════════════════════════

Interceptor.attach(Module.findExportByName(null, "connect"), {
    onEnter(args) {
        try {
            this.fd = args[0].toInt32();
            this.addr = parseSockaddrIn(args[1]);
        } catch (e) {
            this.addr = null;
        }
    },
    onLeave(retval) {
        try {
            const result = retval.toInt32();

            if (this.addr && this.addr.ip) {
                // Enregistrer la connexion
                connections.set(this.fd, {
                    ip: this.addr.ip,
                    port: this.addr.port
                });
                knownSocketFds.add(this.fd);

                const status = result === 0 ? 'OK' : `ERREUR (${result})`;
                const msg = `[CONNECT] fd ${this.fd} → ${this.addr.ip}:${this.addr.port} — ${status}`;
                console.log(`\n${msg}`);

                send({
                    event: 'connect',
                    timestamp_ms: elapsed(),
                    fd: this.fd,
                    ip: this.addr.ip,
                    port: this.addr.port,
                    result: result
                });
            } else if (this.addr) {
                // Famille non AF_INET (AF_UNIX, AF_INET6...)
                console.log(`[CONNECT] fd ${this.fd} — famille ${this.addr.family} (non IPv4)`);
            }
        } catch (e) {
            console.log(`[!] Erreur dans connect/onLeave : ${e.message}`);
        }
    }
});

// ═══════════════════════════════════════════════════════════
// HOOK : send()
// ═══════════════════════════════════════════════════════════

Interceptor.attach(Module.findExportByName(null, "send"), {
    onEnter(args) {
        try {
            this.fd = args[0].toInt32();
            this.buf = args[1];
            this.len = args[2].toInt32();
            this.flags = args[3].toInt32();
            this.shouldLog = shouldLog(this.fd);
        } catch (e) {
            this.shouldLog = false;
            console.log(`[!] Erreur dans send/onEnter : ${e.message}`);
        }
    },
    onLeave(retval) {
        if (!this.shouldLog) return;

        try {
            const bytesSent = retval.toInt32();

            // Incrémenter le numéro de séquence
            sequenceNumber++;
            const seq = sequenceNumber;
            const ts = elapsed();

            // Préparer les métadonnées
            const conn = connections.get(this.fd);
            const meta = {
                event: 'send',
                seq: seq,
                timestamp_ms: ts,
                fd: this.fd,
                requested_len: this.len,
                bytes_sent: bytesSent,
                flags: this.flags,
                dest_ip: conn ? conn.ip : null,
                dest_port: conn ? conn.port : null,
            };

            // Si send() a échoué, logger l'erreur mais ne pas lire le buffer
            if (bytesSent < 0) {
                console.log(`\n[SEND #${seq}] fd ${this.fd} — ERREUR (retour ${bytesSent})`);
                send(meta);
                return;
            }

            // Lire le buffer : au maximum len octets, au maximum bytesSent
            const readSize = Math.min(this.len, bytesSent);
            const bufferData = this.buf.readByteArray(readSize);

            // Affichage console
            const destStr = conn ? ` → ${conn.ip}:${conn.port}` : '';
            console.log(`\n[SEND #${seq}] +${ts}ms | fd ${this.fd}${destStr} | ${bytesSent}/${this.len} octets`);
            console.log(hexdump(this.buf, {
                length: Math.min(readSize, CONFIG.hexdumpMaxBytes)
            }));

            // Envoi vers Python : JSON + buffer binaire brut
            send(meta, bufferData);

        } catch (e) {
            console.log(`[!] Erreur dans send/onLeave : ${e.message}`);
        }
    }
});

// ═══════════════════════════════════════════════════════════
// HOOK OPTIONNEL : write() (pour les sockets)
// ═══════════════════════════════════════════════════════════

if (CONFIG.hookWrite) {
    Interceptor.attach(Module.findExportByName(null, "write"), {
        onEnter(args) {
            try {
                this.fd = args[0].toInt32();
                this.buf = args[1];
                this.len = args[2].toInt32();
                // Ne logger que les write() sur des fd de sockets connus
                this.shouldLog = knownSocketFds.has(this.fd) && shouldLog(this.fd);
            } catch (e) {
                this.shouldLog = false;
            }
        },
        onLeave(retval) {
            if (!this.shouldLog) return;

            try {
                const bytesWritten = retval.toInt32();
                if (bytesWritten < 0) return;

                sequenceNumber++;
                const seq = sequenceNumber;
                const ts = elapsed();
                const readSize = Math.min(this.len, bytesWritten);
                const bufferData = this.buf.readByteArray(readSize);

                const conn = connections.get(this.fd);
                const destStr = conn ? ` → ${conn.ip}:${conn.port}` : '';

                console.log(`\n[WRITE #${seq}] +${ts}ms | fd ${this.fd}${destStr} | ${bytesWritten}/${this.len} octets`);
                console.log(hexdump(this.buf, {
                    length: Math.min(readSize, CONFIG.hexdumpMaxBytes)
                }));

                send({
                    event: 'write',
                    seq: seq,
                    timestamp_ms: ts,
                    fd: this.fd,
                    requested_len: this.len,
                    bytes_sent: bytesWritten,
                    flags: 0,
                    dest_ip: conn ? conn.ip : null,
                    dest_port: conn ? conn.port : null,
                }, bufferData);

            } catch (e) {
                console.log(`[!] Erreur dans write/onLeave : ${e.message}`);
            }
        }
    });
}

// ═══════════════════════════════════════════════════════════
// HOOK : close() — nettoyer les tables de suivi
// ═══════════════════════════════════════════════════════════

Interceptor.attach(Module.findExportByName(null, "close"), {
    onEnter(args) {
        try {
            const fd = args[0].toInt32();
            if (connections.has(fd)) {
                const conn = connections.get(fd);
                console.log(`\n[CLOSE] fd ${fd} (${conn.ip}:${conn.port})`);
                send({
                    event: 'close',
                    timestamp_ms: elapsed(),
                    fd: fd,
                    ip: conn.ip,
                    port: conn.port
                });
                connections.delete(fd);
                knownSocketFds.delete(fd);
            }
        } catch (e) {
            // Ignorer silencieusement — close() est appelé très souvent
        }
    }
});

// ═══════════════════════════════════════════════════════════
// MESSAGE DE DÉMARRAGE
// ═══════════════════════════════════════════════════════════

console.log('═══════════════════════════════════════════════');
console.log('  Frida send() logger — Checkpoint Ch.13');
console.log(`  Filtrage fd : ${CONFIG.filterFd === -1 ? 'désactivé (tout capturer)' : 'fd ' + CONFIG.filterFd}`);
console.log(`  Hook write() : ${CONFIG.hookWrite ? 'oui' : 'non'}`);
console.log('  Ctrl+C pour arrêter');
console.log('═══════════════════════════════════════════════\n');

File: solutions/ch13-checkpoint-solution.py
#!/usr/bin/env python3
"""
solutions/ch13-checkpoint-solution.py

Client Python d'orchestration — Checkpoint Chapitre 13
Spawne le binaire cible, charge l'agent Frida, collecte les
appels à send() et exporte les résultats dans capture.json
et capture.bin.

Usage :
    python3 ch13-checkpoint-solution.py ./binaries/ch13-network/client_O0

    # Avec des arguments pour le binaire cible :
    python3 ch13-checkpoint-solution.py ./binaries/ch13-network/client_O0 -- 127.0.0.1 4444

    # Avec filtrage sur un file descriptor spécifique :
    python3 ch13-checkpoint-solution.py ./binaries/ch13-network/client_O0 --filter-fd 3
"""
import frida
import sys
import os
import json
import time
import argparse
from pathlib import Path

# ═══════════════════════════════════════════════════════════
# CONFIGURATION ET ARGUMENTS
# ═══════════════════════════════════════════════════════════

def parse_args():
    parser = argparse.ArgumentParser(
        description="Frida send() logger — Checkpoint Ch.13"
    )
    parser.add_argument(
        "binary",
        help="Chemin vers le binaire cible"
    )
    parser.add_argument(
        "binary_args",
        nargs="*",
        help="Arguments à passer au binaire cible"
    )
    parser.add_argument(
        "--filter-fd", type=int, default=-1,
        help="Ne logger que ce file descriptor (-1 = tous)"
    )
    parser.add_argument(
        "--output-dir", type=str, default=".",
        help="Répertoire de sortie pour capture.json et capture.bin"
    )
    parser.add_argument(
        "--no-write-hook", action="store_true",
        help="Désactiver le hook sur write()"
    )
    return parser.parse_args()


# ═══════════════════════════════════════════════════════════
# CLASSE PRINCIPALE
# ═══════════════════════════════════════════════════════════

class SendLogger:
    """Orchestre la session Frida et collecte les résultats."""

    def __init__(self, binary, binary_args, filter_fd, output_dir, hook_write):
        self.binary = binary
        self.binary_args = binary_args
        self.filter_fd = filter_fd
        self.output_dir = Path(output_dir)
        self.hook_write = hook_write

        # Données collectées
        self.events = []        # métadonnées JSON de chaque appel
        self.buffers = []       # buffers binaires bruts (bytes)
        self.total_bytes = 0    # octets capturés au total

        # État de la session
        self.session = None
        self.script = None
        self.pid = None
        self.running = True

    def load_agent(self):
        """Charge le code JavaScript de l'agent depuis le fichier."""
        agent_path = Path(__file__).parent / "ch13-checkpoint-solution.js"
        if not agent_path.exists():
            # Fallback : chercher à côté du binaire ou dans le CWD
            agent_path = Path("ch13-checkpoint-solution.js")
        if not agent_path.exists():
            print(f"[!] Agent JS introuvable. Cherché dans :")
            print(f"    - {Path(__file__).parent / 'ch13-checkpoint-solution.js'}")
            print(f"    - {Path('ch13-checkpoint-solution.js').resolve()}")
            sys.exit(1)

        code = agent_path.read_text(encoding="utf-8")

        # Injecter la configuration dans le code agent
        # (remplacer les valeurs par défaut de CONFIG)
        code = code.replace(
            "filterFd: -1,",
            f"filterFd: {self.filter_fd},",
        )
        code = code.replace(
            "hookWrite: true,",
            f"hookWrite: {'true' if self.hook_write else 'false'},",
        )

        return code

    def on_message(self, message, data):
        """
        Callback déclenché par chaque send() côté agent JS.

        - message : dict avec 'type' et 'payload' (le JSON)
        - data    : bytes ou None (le buffer binaire brut)
        """
        if message["type"] == "send":
            payload = message["payload"]
            event_type = payload.get("event")

            if event_type == "connect":
                self._handle_connect(payload)

            elif event_type in ("send", "write"):
                self._handle_send(payload, data)

            elif event_type == "close":
                self._handle_close(payload)

        elif message["type"] == "error":
            print(f"\n[ERREUR AGENT] {message.get('stack', message)}")

    def _handle_connect(self, payload):
        """Affiche et enregistre un événement connect()."""
        result = payload["result"]
        status = "OK" if result == 0 else f"ERREUR ({result})"
        ip = payload.get("ip", "?")
        port = payload.get("port", "?")

        print(f"\n┌─ CONNECT fd {payload['fd']} → {ip}:{port} [{status}]")
        self.events.append(payload)

    def _handle_send(self, payload, data):
        """Affiche et enregistre un événement send()/write()."""
        seq = payload["seq"]
        ts = payload["timestamp_ms"]
        fd = payload["fd"]
        requested = payload["requested_len"]
        sent = payload["bytes_sent"]
        func = payload["event"].upper()
        dest_ip = payload.get("dest_ip")
        dest_port = payload.get("dest_port")

        dest = f" → {dest_ip}:{dest_port}" if dest_ip else ""
        status = f"{sent}/{requested} octets" if sent >= 0 else "ERREUR"

        print(f"│ [{func} #{seq:03d}] +{ts:>6d}ms | fd {fd}{dest} | {status}")

        if data:
            # Afficher un aperçu hex des 48 premiers octets
            preview = data[:48]
            hex_str = " ".join(f"{b:02x}" for b in preview)
            ascii_str = "".join(chr(b) if 32 <= b < 127 else "." for b in preview)
            suffix = "..." if len(data) > 48 else ""
            print(f"│   hex: {hex_str}{suffix}")
            print(f"│   asc: {ascii_str}{suffix}")

            self.buffers.append(data)
            self.total_bytes += len(data)

        # Enregistrer les métadonnées (sans le buffer binaire)
        payload_copy = dict(payload)
        payload_copy["buffer_size"] = len(data) if data else 0
        self.events.append(payload_copy)

    def _handle_close(self, payload):
        """Affiche et enregistre un événement close()."""
        ip = payload.get("ip", "?")
        port = payload.get("port", "?")
        print(f"└─ CLOSE fd {payload['fd']} ({ip}:{port})")
        self.events.append(payload)

    def on_detached(self, reason):
        """Callback quand le processus se termine."""
        print(f"\n[*] Session détachée : {reason}")
        self.running = False

    def export_results(self):
        """Sauvegarde capture.json et capture.bin."""
        self.output_dir.mkdir(parents=True, exist_ok=True)

        # ── capture.json ──
        json_path = self.output_dir / "capture.json"
        export_data = {
            "binary": self.binary,
            "binary_args": self.binary_args,
            "filter_fd": self.filter_fd,
            "total_send_calls": sum(
                1 for e in self.events if e.get("event") in ("send", "write")
            ),
            "total_bytes_captured": self.total_bytes,
            "events": self.events,
        }
        with open(json_path, "w", encoding="utf-8") as f:
            json.dump(export_data, f, indent=2, ensure_ascii=False)
        print(f"[+] Métadonnées exportées → {json_path}")

        # ── capture.bin ──
        bin_path = self.output_dir / "capture.bin"
        with open(bin_path, "wb") as f:
            for buf in self.buffers:
                f.write(buf)
        print(f"[+] Buffers bruts exportés → {bin_path} ({self.total_bytes} octets)")

    def run(self):
        """Point d'entrée principal."""
        # Vérifier que le binaire existe
        if not Path(self.binary).exists():
            print(f"[!] Binaire introuvable : {self.binary}")
            sys.exit(1)

        # Charger l'agent
        agent_code = self.load_agent()

        # Spawner le processus
        spawn_args = [self.binary] + self.binary_args
        print(f"[*] Spawn : {' '.join(spawn_args)}")
        self.pid = frida.spawn(spawn_args)
        self.session = frida.attach(self.pid)
        self.session.on("detached", self.on_detached)

        # Charger le script agent
        self.script = self.session.create_script(agent_code)
        self.script.on("message", self.on_message)
        self.script.load()

        # Reprendre l'exécution du processus
        frida.resume(self.pid)
        print(f"[*] Processus lancé (PID {self.pid}). Ctrl+C pour arrêter.\n")

        # Boucle d'attente
        try:
            while self.running:
                time.sleep(0.5)
        except KeyboardInterrupt:
            print("\n[*] Interruption utilisateur.")

        # Nettoyage et export
        print(f"\n{'═' * 50}")
        print(f"  Résumé de la capture")
        print(f"{'═' * 50}")

        send_count = sum(
            1 for e in self.events if e.get("event") in ("send", "write")
        )
        connect_count = sum(
            1 for e in self.events if e.get("event") == "connect"
        )
        print(f"  Connexions capturées  : {connect_count}")
        print(f"  Appels send()/write() : {send_count}")
        print(f"  Octets totaux         : {self.total_bytes}")
        print(f"{'═' * 50}\n")

        self.export_results()

        # Détacher proprement
        try:
            self.session.detach()
        except Exception:
            pass


# ═══════════════════════════════════════════════════════════
# POINT D'ENTRÉE
# ═══════════════════════════════════════════════════════════

def main():
    args = parse_args()

    logger = SendLogger(
        binary=args.binary,
        binary_args=args.binary_args,
        filter_fd=args.filter_fd,
        output_dir=args.output_dir,
        hook_write=not args.no_write_hook,
    )
    logger.run()


if __name__ == "__main__":
    main()
